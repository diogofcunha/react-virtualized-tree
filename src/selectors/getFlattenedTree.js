export const isNodeExpanded = node => node.state && node.state.expanded;
export const nodeHasChildren = node => node.children && node.children.length;

export const getFlattenedTree = (nodes, parents = []) =>
  nodes.reduce((flattenedTree, node) => {
    const deepness = parents.length;
    const nodeWithHelpers = {...node, deepness, parents};

    if (!nodeHasChildren(node) || !isNodeExpanded(node)) {
      return [...flattenedTree, nodeWithHelpers];
    }

    return [...flattenedTree, nodeWithHelpers, ...getFlattenedTree(node.children, [...parents, node.id])];
  }, []);

export const getFlattenedTreePaths = (nodes, parents = []) => {
  const paths = [];

  for (const node of nodes) {
    const {id} = node;

    if (!nodeHasChildren(node) || !isNodeExpanded(node)) {
      paths.push(parents.concat(id));
    } else {
      paths.push(parents.concat(id));
      paths.push(...getFlattenedTreePaths(node.children, [...parents, id]));
    }
  }

  return paths;
};

export const doesChangeAffectFlattenedTree = (previousNode, nextNode) => {
  return isNodeExpanded(previousNode) !== isNodeExpanded(nextNode);
};
