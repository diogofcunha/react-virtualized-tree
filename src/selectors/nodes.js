import { createSelector } from 'reselect';
import { omit } from 'lodash';

import { UPDATE_TYPE } from '../contants';

export const getNodeRenderOptions = createSelector(
  (node => (node.state || {}).expanded),
  (node => (node.state || {}).favorite),
  (node => (node.state || {}).deletable),
  (node => node.children), 
  (expanded, favorite, deletable, children = []) => ({
    hasChildren: !!children.length,
    isExpanded: !!expanded,
    isFavorite: !!favorite,
    isDeletable: !!deletable
  })
);

const FLATTEN_TREE_PROPERTIES = [ 'deepness', 'parents' ];
const NODE_OPERATION_PROPERTIES = [ 'index' ];

const NODE_OPERATION_TYPES = {
  CHANGE_NODE: 'CHANGE_NODE',
  DELETE_NODE: 'DELETE_NODE',
  ADD_NODE: 'ADD_NODE'
}

const NODE_CHANGE_OPERATIONS = {
  CHANGE_NODE: (nodes, updatedNode) => nodes.map(n => n.id === updatedNode.id ? omit(updatedNode, FLATTEN_TREE_PROPERTIES) : n),
  DELETE_NODE: (nodes, updatedNode) => nodes.filter(n => n.id !== updatedNode.id),
  ADD_NODE: (nodes, nodeToAdd) => [ ...nodes, omit(nodeToAdd, [ ...FLATTEN_TREE_PROPERTIES, ...NODE_OPERATION_PROPERTIES] ) ]
}

export const replaceNodeFromTree = (nodes, updatedNode, operation = NODE_OPERATION_TYPES.CHANGE_NODE) => {
  if (!NODE_CHANGE_OPERATIONS[operation]) {
    return nodes;
  }
  
  const { parents } = updatedNode;

  if (!parents.length) {
    return NODE_CHANGE_OPERATIONS[operation](nodes, updatedNode);
  }

  const parentIndex = nodes.findIndex(n => n.id === parents[0])

  const preSiblings = nodes.slice(0, parentIndex);
  const postSiblings = nodes.slice(parentIndex + 1);

  return [
    ...preSiblings,
    {
      ...nodes[parentIndex], 
      ...nodes[parentIndex].children || operation === NODE_OPERATION_TYPES.ADD_NODE ? 
        { children: replaceNodeFromTree(
            nodes[parentIndex].children || [],
            { ...updatedNode, parents: parents.slice(1) },
            operation
          )
        } : {}
    },
    ...postSiblings
  ]
}

export const deleteNodeFromTree = (nodes, deletedNode) => {
  return replaceNodeFromTree(
    nodes,
    deletedNode,
    NODE_OPERATION_TYPES.DELETE_NODE
  );
}

export const addNodeToTree = (nodes, nodeToAdd) => replaceNodeFromTree(
  nodes,
  nodeToAdd,
  NODE_OPERATION_TYPES.ADD_NODE
)

export const moveNodeToChildren = (nodes, [ fromNode, toNode ]) => {
  const { id: targetId, parents: targetParents } = toNode;
  const parents = [ ...targetParents, ...targetId !== fromNode.id ? [targetId] : []];

  return moveNodeFromTree(nodes, [ fromNode, { ...toNode, parents } ]);
}

export const moveNodeFromTree = (nodes, [ fromNode, toNode ]) => {
  const treeWithoutMovedNode = deleteNodeFromTree(
    nodes,
    fromNode
  );

  return addNodeToTree(treeWithoutMovedNode, { ...fromNode, parents: toNode.parents });
}

export const udpateNode = (originalNode, newState) => ({
  node: {
    ...originalNode,
    state : {
      ...originalNode.state,
      ...newState
    }
  },
  type: UPDATE_TYPE.UPDATE
});

export const deleteNode = node => ({
  node,
  type: UPDATE_TYPE.DELETE
});

export const addNode = node => ({
  node,
  type: UPDATE_TYPE.ADD
});
