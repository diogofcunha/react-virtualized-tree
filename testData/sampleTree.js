export const COLLAPSED_CHILDREN = {
  id: 3,
  name: 'Leaf 3',
  state: {
    expanded: false,
    favorite: true,
    deletable: true,
  },
  children: [
    {
      id: 'c-3',
      name: 'Leaf 3 Child',
      state: {},
    },
  ],
};

export const EXPANDED_CHILDREN = {
  id: 2,
  name: 'Leaf 2',
  state: {
    expanded: true,
    deletable: true,
  },
  children: [
    COLLAPSED_CHILDREN,
    {
      id: 4,
      name: 'Leaf 4',
    },
  ],
};

export const EXPANDED_NODE_IN_ROOT = {
  id: 0,
  name: 'Leaf 1',
  state: {
    expanded: true,
  },
  children: [
    EXPANDED_CHILDREN,
    {
      id: 5,
      name: 'Leaf 5',
    },
  ],
};

export const COLLAPSED_NODE_IN_ROOT = {
  id: 1,
  name: 'Leaf 6',
  state: {
    expanded: false,
    deletable: true,
  },
  children: [
    {
      id: 6,
      name: 'Leaf 7',
      state: {
        expanded: false,
      },
      children: [
        {
          id: 7,
          name: 'Leaf 8',
        },
        {
          id: 8,
          name: 'Leaf 9',
        },
      ],
    },
    {
      id: 9,
      name: 'Leaf 10',
    },
  ],
};

export const DELETABLE_IN_ROOT = {
  id: 'z',
  name: 'Leaf z',
  state: {
    deletable: true,
    favorite: true,
  },
};

export const DELETABLE_CHILDREN = EXPANDED_CHILDREN;

export const Nodes = [
  EXPANDED_NODE_IN_ROOT,
  COLLAPSED_NODE_IN_ROOT,
  DELETABLE_IN_ROOT,
  EXPANDED_NODE_IN_ROOT,
  COLLAPSED_NODE_IN_ROOT,
  DELETABLE_IN_ROOT,
  EXPANDED_NODE_IN_ROOT,
  COLLAPSED_NODE_IN_ROOT,
  DELETABLE_IN_ROOT,
  EXPANDED_NODE_IN_ROOT,
  COLLAPSED_NODE_IN_ROOT,
  DELETABLE_IN_ROOT,
  EXPANDED_NODE_IN_ROOT,
  COLLAPSED_NODE_IN_ROOT,
  DELETABLE_IN_ROOT,
  EXPANDED_NODE_IN_ROOT,
  COLLAPSED_NODE_IN_ROOT,
  DELETABLE_IN_ROOT,
  EXPANDED_NODE_IN_ROOT,
  COLLAPSED_NODE_IN_ROOT,
  DELETABLE_IN_ROOT,
];
