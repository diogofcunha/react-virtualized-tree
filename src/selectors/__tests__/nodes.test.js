import deepFreeze from 'deep-freeze';

import * as nodeSelectors from '../nodes';
import { Nodes } from '../../../testData/sampleTree';
import { getFlattenedTree } from '../getFlattenedTree';

describe('selectors -> nodes ->', () => {
  const getSampleNode = (i = 0) => getFlattenedTree(Nodes)[i];

  describe('single node actions', () => {
    it('updateNode should update the supplied node state without mutations', () => {
      const originalNode = getSampleNode();

      deepFreeze(originalNode);

      expect(
        nodeSelectors.udpateNode(originalNode, { favorite: true, expanded: false } )
      ).toMatchSnapshot();
    });

    it('addNode should create the expected object', () => {
      expect(
        nodeSelectors.addNode(getSampleNode())
      ).toMatchSnapshot();
    });

    it('deleteNode should create the expected object', () => {
      expect(
        nodeSelectors.deleteNode(getSampleNode())
      ).toMatchSnapshot();
    });
  });

  describe('Tree actions', () => {
    describe('deleteNodeFromTree', () => {
      it('when root node is deleted, should delete that node and all children nodes', () => {
        deepFreeze(Nodes);

        expect(
          nodeSelectors.deleteNodeFromTree(Nodes, getSampleNode())
        ).toMatchSnapshot();
      });

      it('when a node is deleted, should delete that node and all children nodes', () => {
        deepFreeze(Nodes);

        expect(
          nodeSelectors.deleteNodeFromTree(Nodes, getSampleNode(1))
        ).toMatchSnapshot();
      });
    });

    it('replaceNodeFromTree should replace a node in the tree without mutations', () => {
      deepFreeze(Nodes);

      const updatedNode =  nodeSelectors.udpateNode(
        getSampleNode(),
        { favorite: true, expanded: false }
      );

      expect(
        nodeSelectors.replaceNodeFromTree(Nodes, updatedNode.node)
      ).toMatchSnapshot();
    });

    describe('moveNodeToChildren', () => {
      it('should move a node from the root correctly without mutations', () => {
        deepFreeze(Nodes);

        const movedNodes = [
          getSampleNode(6), // node z
          getSampleNode(0) // node 1
        ];

        expect(
          nodeSelectors.moveNodeToChildren(Nodes, movedNodes)
        ).toMatchSnapshot();
      });

      it('should move a node to a parentless node without mutations', () => {
        deepFreeze(Nodes);

        const movedNodes = [
          getSampleNode(2), // node 3
          getSampleNode(6) // node z
        ];

        expect(
          nodeSelectors.moveNodeToChildren(Nodes, movedNodes)
        ).toMatchSnapshot();
      });

      it('should move a node to with children without mutations', () => {
        deepFreeze(Nodes);

        const movedNodes = [
          getSampleNode(1), // node 2
          getSampleNode(5) // node 6
        ];

        expect(
          nodeSelectors.moveNodeToChildren(Nodes, movedNodes)
        ).toMatchSnapshot();
      });
    });

    describe('moveNodeFromTree', () => {
      it('should move a node from the root correctly without mutations', () => {
        deepFreeze(Nodes);

        const movedNodes = [
          getSampleNode(6), // node z
          getSampleNode(0) // node 1
        ];

        expect(
          nodeSelectors.moveNodeFromTree(Nodes, movedNodes)
        ).toMatchSnapshot();
      });

      it('should move a node to a parentless node without mutations', () => {
        deepFreeze(Nodes);

        const movedNodes = [
          getSampleNode(2), // node 3
          getSampleNode(6) // node z
        ];

        expect(
          nodeSelectors.moveNodeFromTree(Nodes, movedNodes)
        ).toMatchSnapshot();
      });

      it('should move a node to with children without mutations', () => {
        deepFreeze(Nodes);

        const movedNodes = [
          getSampleNode(1), // node 2
          getSampleNode(5) // node 6
        ];

        expect(
          nodeSelectors.moveNodeFromTree(Nodes, movedNodes)
        ).toMatchSnapshot();
      });
    });
  });

  describe('getNodeRenderOptions', () => {
    it('should extract state from nodes correctly when there are no children', () => {
      expect(
        nodeSelectors.getNodeRenderOptions({ state: { expanded: true, favorite: true } })
      ).toMatchSnapshot();
    });

    it('should extract state from nodes correctly when there are children', () => {
      expect(
        nodeSelectors.getNodeRenderOptions({ children: [{}], state: { deletable: true } })
      ).toMatchSnapshot();
    });

    it('should extract state from nodes correctly when there is no state', () => {
      expect(
        nodeSelectors.getNodeRenderOptions({ children: [{}] })
      ).toMatchSnapshot();
    });
  });
});
