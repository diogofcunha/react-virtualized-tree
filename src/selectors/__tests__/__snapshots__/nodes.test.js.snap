// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`selectors -> nodes -> Tree actions deleteNodeFromTree when a node is deleted, should delete that node and all children nodes 1`] = `
Array [
  Object {
    "children": Array [
      Object {
        "id": 5,
        "name": "Leaf 5",
      },
    ],
    "id": 0,
    "name": "Leaf 1",
    "state": Object {
      "expanded": true,
    },
  },
  Object {
    "children": Array [
      Object {
        "children": Array [
          Object {
            "id": 7,
            "name": "Leaf 8",
          },
          Object {
            "id": 8,
            "name": "Leaf 9",
          },
        ],
        "id": 6,
        "name": "Leaf 7",
        "state": Object {
          "expanded": false,
        },
      },
      Object {
        "id": 9,
        "name": "Leaf 10",
      },
    ],
    "id": 1,
    "name": "Leaf 6",
    "state": Object {
      "deletable": true,
      "expanded": false,
    },
  },
  Object {
    "id": "z",
    "name": "Leaf z",
    "state": Object {
      "deletable": true,
      "favorite": true,
    },
  },
]
`;

exports[`selectors -> nodes -> Tree actions deleteNodeFromTree when root node is deleted, should delete that node and all children nodes 1`] = `
Array [
  Object {
    "children": Array [
      Object {
        "children": Array [
          Object {
            "id": 7,
            "name": "Leaf 8",
          },
          Object {
            "id": 8,
            "name": "Leaf 9",
          },
        ],
        "id": 6,
        "name": "Leaf 7",
        "state": Object {
          "expanded": false,
        },
      },
      Object {
        "id": 9,
        "name": "Leaf 10",
      },
    ],
    "id": 1,
    "name": "Leaf 6",
    "state": Object {
      "deletable": true,
      "expanded": false,
    },
  },
  Object {
    "id": "z",
    "name": "Leaf z",
    "state": Object {
      "deletable": true,
      "favorite": true,
    },
  },
]
`;

exports[`selectors -> nodes -> Tree actions moveNodeFromTree should move a node from the root correctly without mutations 1`] = `
Array [
  Object {
    "children": Array [
      Object {
        "children": Array [
          Object {
            "id": 3,
            "name": "Leaf 3",
            "state": Object {
              "deletable": true,
              "favorite": true,
            },
          },
          Object {
            "id": 4,
            "name": "Leaf 4",
          },
        ],
        "id": 2,
        "name": "Leaf 2",
        "state": Object {
          "deletable": true,
          "expanded": true,
        },
      },
      Object {
        "id": 5,
        "name": "Leaf 5",
      },
    ],
    "id": 0,
    "name": "Leaf 1",
    "state": Object {
      "expanded": true,
    },
  },
  Object {
    "children": Array [
      Object {
        "children": Array [
          Object {
            "id": 7,
            "name": "Leaf 8",
          },
          Object {
            "id": 8,
            "name": "Leaf 9",
          },
        ],
        "id": 6,
        "name": "Leaf 7",
        "state": Object {
          "expanded": false,
        },
      },
      Object {
        "id": 9,
        "name": "Leaf 10",
      },
    ],
    "id": 1,
    "name": "Leaf 6",
    "state": Object {
      "deletable": true,
      "expanded": false,
    },
  },
  Object {
    "id": "z",
    "name": "Leaf z",
    "state": Object {
      "deletable": true,
      "favorite": true,
    },
  },
]
`;

exports[`selectors -> nodes -> Tree actions moveNodeFromTree should move a node to a parentless node without mutations 1`] = `
Array [
  Object {
    "children": Array [
      Object {
        "children": Array [
          Object {
            "id": 4,
            "name": "Leaf 4",
          },
        ],
        "id": 2,
        "name": "Leaf 2",
        "state": Object {
          "deletable": true,
          "expanded": true,
        },
      },
      Object {
        "id": 5,
        "name": "Leaf 5",
      },
    ],
    "id": 0,
    "name": "Leaf 1",
    "state": Object {
      "expanded": true,
    },
  },
  Object {
    "children": Array [
      Object {
        "children": Array [
          Object {
            "id": 7,
            "name": "Leaf 8",
          },
          Object {
            "id": 8,
            "name": "Leaf 9",
          },
        ],
        "id": 6,
        "name": "Leaf 7",
        "state": Object {
          "expanded": false,
        },
      },
      Object {
        "id": 9,
        "name": "Leaf 10",
      },
    ],
    "id": 1,
    "name": "Leaf 6",
    "state": Object {
      "deletable": true,
      "expanded": false,
    },
  },
  Object {
    "id": "z",
    "name": "Leaf z",
    "state": Object {
      "deletable": true,
      "favorite": true,
    },
  },
  Object {
    "id": 3,
    "name": "Leaf 3",
    "state": Object {
      "deletable": true,
      "favorite": true,
    },
  },
]
`;

exports[`selectors -> nodes -> Tree actions moveNodeFromTree should move a node to with children without mutations 1`] = `
Array [
  Object {
    "children": Array [
      Object {
        "id": 5,
        "name": "Leaf 5",
      },
    ],
    "id": 0,
    "name": "Leaf 1",
    "state": Object {
      "expanded": true,
    },
  },
  Object {
    "children": Array [
      Object {
        "children": Array [
          Object {
            "id": 7,
            "name": "Leaf 8",
          },
          Object {
            "id": 8,
            "name": "Leaf 9",
          },
        ],
        "id": 6,
        "name": "Leaf 7",
        "state": Object {
          "expanded": false,
        },
      },
      Object {
        "id": 9,
        "name": "Leaf 10",
      },
    ],
    "id": 1,
    "name": "Leaf 6",
    "state": Object {
      "deletable": true,
      "expanded": false,
    },
  },
  Object {
    "id": "z",
    "name": "Leaf z",
    "state": Object {
      "deletable": true,
      "favorite": true,
    },
  },
  Object {
    "children": Array [
      Object {
        "id": 3,
        "name": "Leaf 3",
        "state": Object {
          "deletable": true,
          "favorite": true,
        },
      },
      Object {
        "id": 4,
        "name": "Leaf 4",
      },
    ],
    "id": 2,
    "name": "Leaf 2",
    "state": Object {
      "deletable": true,
      "expanded": true,
    },
  },
]
`;

exports[`selectors -> nodes -> Tree actions moveNodeToChildren should move a node from the root correctly without mutations 1`] = `
Array [
  Object {
    "children": Array [
      Object {
        "children": Array [
          Object {
            "id": 3,
            "name": "Leaf 3",
            "state": Object {
              "deletable": true,
              "favorite": true,
            },
          },
          Object {
            "id": 4,
            "name": "Leaf 4",
          },
        ],
        "id": 2,
        "name": "Leaf 2",
        "state": Object {
          "deletable": true,
          "expanded": true,
        },
      },
      Object {
        "id": 5,
        "name": "Leaf 5",
      },
      Object {
        "id": "z",
        "name": "Leaf z",
        "state": Object {
          "deletable": true,
          "favorite": true,
        },
      },
    ],
    "id": 0,
    "name": "Leaf 1",
    "state": Object {
      "expanded": true,
    },
  },
  Object {
    "children": Array [
      Object {
        "children": Array [
          Object {
            "id": 7,
            "name": "Leaf 8",
          },
          Object {
            "id": 8,
            "name": "Leaf 9",
          },
        ],
        "id": 6,
        "name": "Leaf 7",
        "state": Object {
          "expanded": false,
        },
      },
      Object {
        "id": 9,
        "name": "Leaf 10",
      },
    ],
    "id": 1,
    "name": "Leaf 6",
    "state": Object {
      "deletable": true,
      "expanded": false,
    },
  },
]
`;

exports[`selectors -> nodes -> Tree actions moveNodeToChildren should move a node to a parentless node without mutations 1`] = `
Array [
  Object {
    "children": Array [
      Object {
        "children": Array [
          Object {
            "id": 4,
            "name": "Leaf 4",
          },
        ],
        "id": 2,
        "name": "Leaf 2",
        "state": Object {
          "deletable": true,
          "expanded": true,
        },
      },
      Object {
        "id": 5,
        "name": "Leaf 5",
      },
    ],
    "id": 0,
    "name": "Leaf 1",
    "state": Object {
      "expanded": true,
    },
  },
  Object {
    "children": Array [
      Object {
        "children": Array [
          Object {
            "id": 7,
            "name": "Leaf 8",
          },
          Object {
            "id": 8,
            "name": "Leaf 9",
          },
        ],
        "id": 6,
        "name": "Leaf 7",
        "state": Object {
          "expanded": false,
        },
      },
      Object {
        "id": 9,
        "name": "Leaf 10",
      },
    ],
    "id": 1,
    "name": "Leaf 6",
    "state": Object {
      "deletable": true,
      "expanded": false,
    },
  },
  Object {
    "children": Array [
      Object {
        "id": 3,
        "name": "Leaf 3",
        "state": Object {
          "deletable": true,
          "favorite": true,
        },
      },
    ],
    "id": "z",
    "name": "Leaf z",
    "state": Object {
      "deletable": true,
      "favorite": true,
    },
  },
]
`;

exports[`selectors -> nodes -> Tree actions moveNodeToChildren should move a node to with children without mutations 1`] = `
Array [
  Object {
    "children": Array [
      Object {
        "id": 5,
        "name": "Leaf 5",
      },
    ],
    "id": 0,
    "name": "Leaf 1",
    "state": Object {
      "expanded": true,
    },
  },
  Object {
    "children": Array [
      Object {
        "children": Array [
          Object {
            "id": 7,
            "name": "Leaf 8",
          },
          Object {
            "id": 8,
            "name": "Leaf 9",
          },
        ],
        "id": 6,
        "name": "Leaf 7",
        "state": Object {
          "expanded": false,
        },
      },
      Object {
        "id": 9,
        "name": "Leaf 10",
      },
      Object {
        "children": Array [
          Object {
            "id": 3,
            "name": "Leaf 3",
            "state": Object {
              "deletable": true,
              "favorite": true,
            },
          },
          Object {
            "id": 4,
            "name": "Leaf 4",
          },
        ],
        "id": 2,
        "name": "Leaf 2",
        "state": Object {
          "deletable": true,
          "expanded": true,
        },
      },
    ],
    "id": 1,
    "name": "Leaf 6",
    "state": Object {
      "deletable": true,
      "expanded": false,
    },
  },
  Object {
    "id": "z",
    "name": "Leaf z",
    "state": Object {
      "deletable": true,
      "favorite": true,
    },
  },
]
`;

exports[`selectors -> nodes -> Tree actions replaceNodeFromTree should replace a node in the tree without mutations 1`] = `
Array [
  Object {
    "children": Array [
      Object {
        "children": Array [
          Object {
            "id": 3,
            "name": "Leaf 3",
            "state": Object {
              "deletable": true,
              "favorite": true,
            },
          },
          Object {
            "id": 4,
            "name": "Leaf 4",
          },
        ],
        "id": 2,
        "name": "Leaf 2",
        "state": Object {
          "deletable": true,
          "expanded": true,
        },
      },
      Object {
        "id": 5,
        "name": "Leaf 5",
      },
    ],
    "id": 0,
    "name": "Leaf 1",
    "state": Object {
      "expanded": false,
      "favorite": true,
    },
  },
  Object {
    "children": Array [
      Object {
        "children": Array [
          Object {
            "id": 7,
            "name": "Leaf 8",
          },
          Object {
            "id": 8,
            "name": "Leaf 9",
          },
        ],
        "id": 6,
        "name": "Leaf 7",
        "state": Object {
          "expanded": false,
        },
      },
      Object {
        "id": 9,
        "name": "Leaf 10",
      },
    ],
    "id": 1,
    "name": "Leaf 6",
    "state": Object {
      "deletable": true,
      "expanded": false,
    },
  },
  Object {
    "id": "z",
    "name": "Leaf z",
    "state": Object {
      "deletable": true,
      "favorite": true,
    },
  },
]
`;

exports[`selectors -> nodes -> getNodeRenderOptions should extract state from nodes correctly when there are children 1`] = `
Object {
  "hasChildren": true,
  "isDeletable": true,
  "isExpanded": false,
  "isFavorite": false,
}
`;

exports[`selectors -> nodes -> getNodeRenderOptions should extract state from nodes correctly when there are no children 1`] = `
Object {
  "hasChildren": false,
  "isDeletable": false,
  "isExpanded": true,
  "isFavorite": true,
}
`;

exports[`selectors -> nodes -> getNodeRenderOptions should extract state from nodes correctly when there is no state 1`] = `
Object {
  "hasChildren": true,
  "isDeletable": false,
  "isExpanded": false,
  "isFavorite": false,
}
`;

exports[`selectors -> nodes -> single node actions addNode should create the expected object 1`] = `
Object {
  "node": Object {
    "children": Array [
      Object {
        "children": Array [
          Object {
            "id": 3,
            "name": "Leaf 3",
            "state": Object {
              "deletable": true,
              "favorite": true,
            },
          },
          Object {
            "id": 4,
            "name": "Leaf 4",
          },
        ],
        "id": 2,
        "name": "Leaf 2",
        "state": Object {
          "deletable": true,
          "expanded": true,
        },
      },
      Object {
        "id": 5,
        "name": "Leaf 5",
      },
    ],
    "deepness": 0,
    "id": 0,
    "name": "Leaf 1",
    "parents": Array [],
    "state": Object {
      "expanded": true,
    },
  },
  "type": 0,
}
`;

exports[`selectors -> nodes -> single node actions deleteNode should create the expected object 1`] = `
Object {
  "node": Object {
    "children": Array [
      Object {
        "children": Array [
          Object {
            "id": 3,
            "name": "Leaf 3",
            "state": Object {
              "deletable": true,
              "favorite": true,
            },
          },
          Object {
            "id": 4,
            "name": "Leaf 4",
          },
        ],
        "id": 2,
        "name": "Leaf 2",
        "state": Object {
          "deletable": true,
          "expanded": true,
        },
      },
      Object {
        "id": 5,
        "name": "Leaf 5",
      },
    ],
    "deepness": 0,
    "id": 0,
    "name": "Leaf 1",
    "parents": Array [],
    "state": Object {
      "expanded": true,
    },
  },
  "type": 1,
}
`;

exports[`selectors -> nodes -> single node actions updateNode should update the supplied node state without mutations 1`] = `
Object {
  "node": Object {
    "children": Array [
      Object {
        "children": Array [
          Object {
            "id": 3,
            "name": "Leaf 3",
            "state": Object {
              "deletable": true,
              "favorite": true,
            },
          },
          Object {
            "id": 4,
            "name": "Leaf 4",
          },
        ],
        "id": 2,
        "name": "Leaf 2",
        "state": Object {
          "deletable": true,
          "expanded": true,
        },
      },
      Object {
        "id": 5,
        "name": "Leaf 5",
      },
    ],
    "deepness": 0,
    "id": 0,
    "name": "Leaf 1",
    "parents": Array [],
    "state": Object {
      "expanded": false,
      "favorite": true,
    },
  },
  "type": 2,
}
`;
